File: 14_并查集/WeightedUnionFind-分量和.py
from collections import defaultdict
from typing import Callable, DefaultDict, List, Optional, Tuple


class WeightedUnionFind:
    """维护分量和的并查集."""

    __slots__ = "_parent", "_value", "_delta", "_total", "_part"

    def __init__(self, n: int):
        self._parent = [-1] * n
        self._value = [0] * n
        self._delta = [0] * n
        self._total = [0] * n
        self._part = n

    def add(self, u: int, delta: int) -> None:
        """u的值加上delta."""
        self._value[u] += delta
        self._total[self.find(u)] += delta

    def addGroup(self, u: int, delta: int) -> None:
        """u所在集合的值加上delta."""
        root = self.find(u)
        self._delta[root] += delta
        self._total[root] -= self._parent[root] * delta

    def get(self, u: int) -> int:
        """u的值."""
        return self._value[u] + self._find(u)[1]

    def getGroup(self, u: int) -> int:
        """u所在集合的值."""
        return self._total[self.find(u)]

    def union(self, u: int, v: int, f: Optional[Callable[[int, int], None]] = None) -> bool:
        u = self.find(u)
        v = self.find(v)
        if u == v:
            return False
        if self._parent[u] > self._parent[v]:
            u, v = v, u
        self._parent[u] += self._parent[v]
        self._parent[v] = u
        self._delta[v] -= self._delta[u]
        self._total[u] += self._total[v]
        self._part -= 1
        if f:
            f(u, v)
        return True

    def find(self, u: int) -> int:
        return self._find(u)[0]

    def isConnected(self, u: int, v: int) -> bool:
        return self.find(u) == self.find(v)

    def getSize(self, u: int) -> int:
        return -self._parent[self.find(u)]

    def getGroups(self) -> DefaultDict[int, List[int]]:
        groups = defaultdict(list)
        for i in range(len(self._parent)):
            root = self.find(i)
            if root in groups:
                groups[root].append(i)
            else:
                groups[root] = [i]
        return groups

    @property
    def part(self) -> int:
        return self._part

    def _find(self, u: int) -> Tuple[int, int]:
        if self._parent[u] < 0:
            return u, self._delta[u]
        p = self._find(self._parent[u])
        first = p[0]
        second = p[1] + self._delta[u] - self._delta[first]
        self._parent[u] = first
        self._delta[u] = second
        return p


if __name__ == "__main__":
    N = 5
    uf = WeightedUnionFind(N)
    for i in range(N):
        uf.add(i, i)
    uf.union(2, 4)
    for i in range(N):
        print(uf.get(i), uf.getGroup(i))
    print()
    uf.union(3, 4)
    for i in range(N):
        print(uf.get(i), uf.getGroup(i))
    print()
    uf.add(3, 10)
    for i in range(N):
        print(uf.get(i), uf.getGroup(i))
    print()
    uf.addGroup(4, 5)
    for i in range(N):
        print(uf.get(i), uf.getGroup(i))

File: 14_并查集/WeightedUnionFind-分量和.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	// https://yukicoder.me/problems/no/1054
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)

	uf := NewWeightedUnionFind(n)
	for i := 0; i < q; i++ {
		var op, a, b int
		fmt.Fscan(in, &op, &a, &b)
		if op == 1 {
			a, b = a-1, b-1
			uf.Union(a, b)
		} else if op == 2 {
			a -= 1
			uf.AddGroup(a, b)
		} else {
			a -= 1
			fmt.Fprintln(out, uf.Get(a))
		}
	}

}

// 维护分量和的并查集.
type WeightedUnionFind struct {
	Part   int
	parent []int
	value  []int
	delta  []int
	total  []int
}

// NewUnionFindWeighted
func NewWeightedUnionFind(n int) *WeightedUnionFind {
	uf := &WeightedUnionFind{
		Part:   n,
		parent: make([]int, n),
		value:  make([]int, n),
		delta:  make([]int, n),
		total:  make([]int, n),
	}
	for i := range uf.parent {
		uf.parent[i] = -1
	}
	return uf
}

// u的值加上delta.
func (uf *WeightedUnionFind) Add(u, delta int) {
	uf.value[u] += delta
	uf.total[uf.Find(u)] += delta
}

// u所在集合的值加上delta.
func (uf *WeightedUnionFind) AddGroup(u, delta int) {
	root := uf.Find(u)
	uf.delta[root] += delta
	uf.total[root] -= uf.parent[root] * delta
}

// u的值.
func (uf *WeightedUnionFind) Get(u int) int {
	_, delta := uf._find(u)
	return uf.value[u] + delta
}

// u所在集合的值.
func (uf *WeightedUnionFind) GetGroup(u int) int {
	return uf.total[uf.Find(u)]
}

func (uf *WeightedUnionFind) Union(u, v int) bool {
	u = uf.Find(u)
	v = uf.Find(v)
	if u == v {
		return false
	}
	if uf.parent[u] > uf.parent[v] {
		u ^= v
		v ^= u
		u ^= v
	}
	uf.parent[u] += uf.parent[v]
	uf.parent[v] = u
	uf.delta[v] -= uf.delta[u]
	uf.total[u] += uf.total[v]
	uf.Part--
	return true
}

func (uf *WeightedUnionFind) UnionWithCallback(u, v int, f func(u, v int)) bool {
	u = uf.Find(u)
	v = uf.Find(v)
	if u == v {
		return false
	}
	if uf.parent[u] > uf.parent[v] {
		u ^= v
		v ^= u
		u ^= v
	}
	uf.parent[u] += uf.parent[v]
	uf.parent[v] = u
	uf.delta[v] -= uf.delta[u]
	uf.total[u] += uf.total[v]
	uf.Part--
	if f != nil {
		f(u, v)
	}
	return true
}

// u的根.
func (uf *WeightedUnionFind) Find(u int) int {
	root, _ := uf._find(u)
	return root
}

// u和v是否连通.
func (uf *WeightedUnionFind) IsConnected(u, v int) bool {
	return uf.Find(u) == uf.Find(v)
}

// u所在集合的大小.
func (uf *WeightedUnionFind) GetSize(u int) int {
	return -uf.parent[uf.Find(u)]
}

// 所有分量.
func (uf *WeightedUnionFind) GetGroups() map[int][]int {
	groups := make(map[int][]int)
	for i := range uf.parent {
		root := uf.Find(i)
		groups[root] = append(groups[root], i)
	}
	return groups
}

func (uf *WeightedUnionFind) _find(u int) (root, delta int) {
	if uf.parent[u] < 0 {
		return u, uf.delta[u]
	}
	root, delta = uf._find(uf.parent[u])
	delta += uf.delta[u]
	uf.parent[u] = root
	uf.delta[u] = delta - uf.delta[root]
	return
}
